src/app/settings/admin/constraint-templates/component.ts

```diff
@@ -0,0 +1,156 @@
+import {MatTableDataSource} from '@angular/material/table';
+import {OPAService} from '@core/services/opa/service';
+import {NotificationService} from '@core/services/notification/service';
+import {UserService} from '@core/services/user/service';
+import {ConstraintTemplate} from '@shared/entity/opa';
+import * as _ from 'lodash';
+import {Subject} from 'rxjs';
+import {takeUntil, filter, take, switchMap} from 'rxjs/operators';
+import {ConstraintTemplateDialog} from './constraint-template-dialog/component';
+import {DeleteConstraintTemplateDialog} from './delete-constraint-template-dialog/component';
+
+@Component({
+  selector: 'km-constraint-templates-list',
+  templateUrl: './template.html',
+})
+export class ConstraintTemplatesComponent implements OnInit, OnChanges, OnDestroy {
+  constraintTemplates: ConstraintTemplate[] = [];
+  dataSource = new MatTableDataSource<ConstraintTemplate>();
+  displayedColumns: string[] = ['templateName', 'actions'];
+  @ViewChild(MatSort, {static: true}) sort: MatSort;
+  @ViewChild(MatPaginator, {static: true}) paginator: MatPaginator;
+  private readonly _unsubscribe = new Subject<void>();
+  private readonly _defaultTimeout = 3000;
+
+  constructor(
+    private readonly _opaService: OPAService,
+    private readonly _userService: UserService,
+    private readonly _notificationService: NotificationService,
+    private readonly _matDialog: MatDialog
+  ) {}
+
+  ngOnInit() {
+    this.dataSource.data = this.constraintTemplates;
+    this.dataSource.sort = this.sort;
+    this.dataSource.paginator = this.paginator;
+    this.sort.active = 'templateName';
+    this.sort.direction = 'asc';
+
+    this._opaService.constraintTemplates.pipe(takeUntil(this._unsubscribe)).subscribe(constraintTemplates => {
+      this.constraintTemplates = constraintTemplates;
+      this.dataSource.data = this.constraintTemplates;
+    });
+
+    this._userService.currentUserSettings.pipe(takeUntil(this._unsubscribe)).subscribe(settings => {
+      this.paginator.pageSize = settings.itemsPerPage;
+      this.dataSource.paginator = this.paginator; // Force refresh.
+    });
+  }
+
+  ngOnChanges(): void {
+    this.dataSource.data = this.constraintTemplates;
+  }
+
+  ngOnDestroy(): void {
+    this._unsubscribe.next();
+    this._unsubscribe.complete();
+  }
+
+  isPaginatorVisible(): boolean {
+    return (
+      this.constraintTemplates &&
+      this.constraintTemplates.length > 0 &&
+      this.paginator &&
+      this.constraintTemplates.length > this.paginator.pageSize
+    );
+  }
+
+  add(): void {
+    const dialogConfig: MatDialogConfig = {
+      data: {
+        title: 'Add Constraint Template',
+        isEditing: false,
+        confirmLabel: 'Add',
+      },
+    };
+
+    this._matDialog
+      .open(ConstraintTemplateDialog, dialogConfig)
+      .afterClosed()
+      .pipe(filter(constraintTemplate => !!constraintTemplate))
+      .pipe(take(1))
+      .subscribe((result: ConstraintTemplate) => {
```

Comment:
nit: For 1 line arrow functions there is no need to add additional brackets. It can be inlined.

Reply:
You can use `bind` method here also instead of calling private add method inside arrow function.

Reply:
Shouldn't actually the dialog take care of creating/editing here instead of only acting as a proxy for data? @maciaszczykm WDYT?

Reply:
It's hard to say at a first glance. I remember when I was implementing addons I had to take the logic outside of the addons component because it will act different inside cluster page and inside wizard (wizard part is not implemented yet). So, I'd say there might be some cases where we can take the logic outside and it makes sense.

However, I'm not sure if we have any pattern how to do it in general. I mean keeping logic in the components vs in the dialogs. I think in most of the cases we keep it in the dialogs so that might be a good point to stick to that.

Reply:
Alright, changed it in https://github.com/kubermatic/dashboard/pull/2999/commits/ad556c25a3ffe1aa3f96addc434711da4df036fe